peer-ids/peer-ids.md

```diff
@@ -0,0 +1,117 @@
+# Spec:  Peer Ids and Keys
+
+## Keys
+
+
+Our key pairs are stored on disk using a simple protobuf defined in [libp2p/go-libp2p-crypto/pb/crypto.proto#L5](https://github.com/libp2p/go-libp2p-crypto/blob/master/pb/crypto.proto#L5):
+
+```protobuf
+enum KeyType {
+	RSA = 0;
+	Ed25519 = 1;
+	Secp256k1 = 2;
+	ECDSA = 3;
+}
+
+message PublicKey {
+	required KeyType Type = 1;
+	required bytes Data = 2;
+}
+
+message PrivateKey {
+	required KeyType Type = 1;
+	required bytes Data = 2;
+}
+```
+
+As should be apparent from the above code block, this proto simply encodes for transmission a public/private key pair along with an enum specifying the type of keypair.
```

Comment:
Is there any situation where we want to transmit the `PrivateKey`? That seems... dangerous. If not, we don't need to specify the `PrivateKey` here at all.

Reply:
Yeah, storage of private key is implementation specific, so no need to cover them in this doc I think.

Reply:
Unfortunately, users do need to be able to take their private keys with them (especially because we use these for things like IPNS).

Reply:
It's true that removing the private key format from this doc leaves a gap. We still need to specify somewhere how we handle them.

We could bring back the private key references and add a call-out at the top of the doc that they're not related to peer-id calculation and are shown for reference.

Reply:
Really, we should probably rename this doc to the "libp2p key spec" and make peer ID calculation a part of that.

Reply:
👍 for that